example/js/js_image_layer.bzl

```diff
@@ -0,0 +1,260 @@
+    files = depset([executable, launcher], transitive = [files])
+
+    symlinks = []
+
+    # NOTE: symlinks is different than root_symlinks. See: https://bazel.build/rules/rules#runfiles_symlinks for distinction between
+    # root_symlinks and symlinks and why they have to be handled differently.
+    for symlink in default.data_runfiles.symlinks.to_list():
+        destination = "/".join([runfiles_dir, ctx.workspace_name, symlink.path])
+        if not _should_include(destination, ctx.attr.include, ctx.attr.exclude):
+            continue
+        if hasattr(file_map, destination):
+            file_map.pop(destination)
+        info = PackageSymlinkInfo(
+            target = "/%s" % _runfile_path(ctx, symlink.target_file, runfiles_dir),
+            destination = destination,
+            attributes = {"mode": "0777"},
+        )
+        symlinks.append([info, symlink.target_file.owner])
+
+    for symlink in default.data_runfiles.root_symlinks.to_list():
+        destination = "/".join([runfiles_dir, symlink.path])
+        if not _should_include(destination, ctx.attr.include, ctx.attr.exclude):
+            continue
+        if hasattr(file_map, destination):
+            file_map.pop(destination)
+        info = PackageSymlinkInfo(
+            target = "/%s" % _runfile_path(ctx, symlink.target_file, runfiles_dir),
+            destination = destination,
+            attributes = {"mode": "0777"},
+        )
+        symlinks.append([info, symlink.target_file.owner])
+
+    return [
+        PackageFilegroupInfo(
+            pkg_dirs = [],
+            pkg_files = [
+                [PackageFilesInfo(
+                    dest_src_map = file_map,
+                    attributes = {},
+                ), ctx.label],
+            ],
+            pkg_symlinks = symlinks,
+        ),
+        DefaultInfo(files = files),
+    ]
+
+runfiles = rule(
+    implementation = _runfiles_impl,
+    attrs = {
+        "binary": attr.label(mandatory = True),
+        "root": attr.string(),
+        "include": attr.string(),
+        "exclude": attr.string(),
+    },
+)
+
+# pkg_tar has poor support for symlinks due to bazel not providing enough information about symlinks.
+#
+# ```starlark
+#   actual_file = ctx.actions.declare_file("thefile.txt")
+#   symlink_file = ctx.actions.declare_file("this_is_a_symlink.txt")
+#   ctx.actions.symlink(symlink_file, target_file = actual_file)
+# ```
+# to determine a file is file or directory pkg_tar checks `is_directory` if it is false then it makes the assumption
+# that the file is in fact is a file which not true for `symlink_file` above. this is where pkg_tar fails to build
+# a correct tar file.
+# In order to fix this, manifest written pkg_tar should be fixed by looking for files that are symlink instead of file.
+
+# TODO(thesayyn): remove this once pkg_tar is fixed.
+#
+# See: https://github.com/bazelbuild/rules_pkg/issues/115#issuecomment-1190494335
+BUILD_TAR = """
```

Comment:
I foresee that this will live here for a while still. Since this is not a trivial code, please move it to a separate file and consume it as input to the rule. Ideally, you would also set up a linter for Python. Can be using pre-commit.

Reply:
It doesn't need to be visible to users. Again, this is not trivial code; please, move it to a separate file.

Reply:
still, it's easier to tell users to copy just one file than two and change their build files. left a comment here to remind us to move it to a separate file when we promote it to the public API. https://github.com/aspect-build/rules_js/issues/304#issuecomment-1325935905